### ISSUE TITLE
Reproduced results and eval problems

### ISSUE THREAD

**Issue by Fantasy1120** [TinyLLaVA/TinyLLaVA_Factory#82]
I try to reproduce the results under base recipe. I basically get the results in the paper on VQAv2, GQA, ScienceQA and POPE. But there is almost 1% gap on TextVQA, MMMU, and MM-Vet, and the gap on MME seems to be larger. I'm not sure if this gap is acceptable? Or what could potentially cause this gap?
<img width="820" alt="reproduce" src="https://github.com/TinyLLaVA/TinyLLaVA_Factory/assets/61446184/8a60b31e-d7d4-44ff-a8bf-0d85740a5010">

**Comment by YingHuTsing** [comment]
Hi, I think this is acceptable. Different number of GPUs cause different gradient_accumulation_steps. Different types of GPUs cause randomness. Btw, the performance for phi-2-siglip-base we listed here is trained by 8 A100-40Gs.

**Comment by Fantasy1120** [comment]
> Hi, I think this is acceptable. Different number of GPUs cause different gradient_accumulation_steps. Different types of GPUs cause randomness. Btw, the performance for phi-2-siglip-base we listed here is trained by 8 A100-40Gs.

Thanks for your reply. I see that you are using fp16 by default in the training script, but A100 supports bf16. May I ask if you are using the bf16 in your training?

**Comment by YingHuTsing** [comment]
No, we haven't tried bf16 thoroughly. But we encourage the open-source community to give it a try. And we can update the performance table accordingly and welcome the open-source community as the contributors of this code repository.

**Comment by pumetu** [comment]
Hello, I've been trying to reproduce the results in the paper but have been struggling to get the same results. Would you be able to share the configuration you used when training?

**Comment by Fantasy1120** [comment]
Hi. I found that different seeds have an effect on the results. My reproduction results when using the phi-2 series are as follows:
<img width="770" alt="Snipaste_2024-11-25_17-07-06" src="https://github.com/user-attachments/assets/a772fad1-0b73-4fcb-9d7f-2df5920f3812">

Additionally, I modified the way I loaded the model during the fine-tuning phase.  For me, it seems to be able to reproduce consistently after the modification, hope this helps.

` if training_arguments.pretrained_model_path is not None:
        # model = training_recipe.load(model, model_args)
        # load from vision pretrained
        model.vision_tower._vision_tower.load_state_dict(pre_vision)
        print(f"Loading vision from {pre_vision_tower_path}")

        # load from connector pretrained
        model.connector._connector.load_state_dict(get_conncet_w(pre_connector, '_connector'))
        print(f"Loading connector from {pre_connector_path}")
        
        # load from pretrained
        for key, Value in pre_language.state_dict().items():
            model.language_model.state_dict()[key].copy_(Value)
        for key in pre_language.state_dict().keys():
            assert torch.equal(pre_language.state_dict()[key],
                            model.language_model.state_dict()[key]), f"Mismatch found in parameter: {key}"
        print(f"Loading language from Huggingface Pretrained Model")
` where 
` 
        pre_vision = torch.load(os.path.join('./pretrain_ckpt', 'vision_tower/pytorch_model.bin'), map_location='cpu')
        pre_connector = torch.load(os.path.join('./pretrain_ckpt', 'connector/pytorch_model.bin'), map_location='cpu')
        pre_language =  AutoModelForCausalLM.from_pretrained(os.path.join('./huggingface_ckpt', 'Qwen1.5-XXB'), trust_remote_code=True)
`

**Comment by Fantasy1120** [comment]
> Hello, I've been trying to reproduce the results in the paper but have been struggling to get the same results. Would you be able to share the configuration you used when training?
> 
> Here is my current config
> 
> ```shell
> DATA_PATH='datasets/multimodal/train_json/sharegpt4v_1246k.json'
> IMAGE_PATH='datasets/multimodal'
> LLM_VERSION='models/phi-2'
> VT_VERSION='models/siglip-so400m-patch14-384'
> VT_VERSION2=""
> CN_VERSION="mlp2x_gelu"
> TRAIN_RECIPE="common"
> MODEL_MAX_LENGTH="3072"
> OUTPUT_DIR='./models/pretrain/phi2-2.7B-siglip-pretrain'
> RUN_NAME='tinyllava-phi2-2.7B-siglip-pretrain'
> 
> deepspeed tinyllava/train/train.py \
>     --deepspeed ./scripts/zero3.json \
>     --data_path  $DATA_PATH\
>     --image_folder $IMAGE_PATH \
>     --is_multimodal True \
>     --conv_version pretrain \
>     --model_name_or_path $LLM_VERSION \
>     --vision_tower $VT_VERSION \
>     --vision_tower2 "$VT_VERSION2" \
>     --connector_type $CN_VERSION \
>     --mm_vision_select_layer -2 \
>     --image_aspect_ratio square \
>     --attn_implementation flash_attention_2 \
>     --fp16 True \
>     --training_recipe $TRAIN_RECIPE \
>     --tune_type_llm frozen \
>     --tune_type_vision_tower frozen \
>     --tune_vision_tower_from_layer 0 \
>     --tune_type_connector full \
>     --output_dir $OUTPUT_DIR \
>     --num_train_epochs 1 \
>     --per_device_train_batch_size 16 \
>     --per_device_eval_batch_size 4 \
>     --gradient_accumulation_steps 4 \
>     --evaluation_strategy "no" \
>     --save_strategy "steps" \
>     --save_steps 24000 \
>     --save_total_limit 1 \
>     --learning_rate 1e-3 \
>     --weight_decay 0. \
>     --warmup_ratio 0.03 \
>     --lr_scheduler_type "cosine" \
>     --logging_steps 1 \
>     --tf32 False \
>     --model_max_length $MODEL_MAX_LENGTH \
>     --gradient_checkpointing True \
>     --dataloader_num_workers 8 \
>     --lazy_preprocess True \
>     --report_to wandb \
>     --tokenizer_use_fast False \
>     --run_name $RUN_NAME
> ```

Hi. I found that different seeds have an effect on the results. My reproduction results when using the phi-2 series are as follows:
<img width="770" alt="Snipaste_2024-11-25_17-07-06" src="https://github.com/user-attachments/assets/a772fad1-0b73-4fcb-9d7f-2df5920f3812">

Additionally, I modified the way I loaded the model during the fine-tuning phase.  For me, it seems to be able to reproduce consistently after the modification, hope this helps.

` 
        pre_vision = torch.load(os.path.join('./pretrain_ckpt', 'vision_tower/pytorch_model.bin'), map_location='cpu')
        pre_connector = torch.load(os.path.join('./pretrain_ckpt', 'connector/pytorch_model.bin'), map_location='cpu')
        pre_language =  AutoModelForCausalLM.from_pretrained(os.path.join('./huggingface_ckpt', 'Qwen1.5-XXB'), trust_remote_code=True)`
        
if training_arguments.pretrained_model_path is not None:
        # model = training_recipe.load(model, model_args)
        # load from vision pretrained
        model.vision_tower._vision_tower.load_state_dict(pre_vision)
        print(f"Loading vision from {pre_vision_tower_path}")

        # load from connector pretrained
        model.connector._connector.load_state_dict(get_conncet_w(pre_connector, '_connector'))
        print(f"Loading connector from {pre_connector_path}")
        
        # load from pretrained
        for key, Value in pre_language.state_dict().items():
            model.language_model.state_dict()[key].copy_(Value)
        for key in pre_language.state_dict().keys():
            assert torch.equal(pre_language.state_dict()[key],
                            model.language_model.state_dict()[key]), f"Mismatch found in parameter: {key}"
        print(f"Loading language from Huggingface Pretrained Model")
`
